crypto/pkcs7/pk7_doit.c

```diff
@@ -346,8 +383,18 @@ BIO *PKCS7_dataInit(PKCS7 *p7, BIO *bio)
     if (bio == NULL) {
         if (PKCS7_is_detached(p7)) {
             bio = BIO_new(BIO_s_null());
-        } else if (os && os->length > 0) {
-            bio = BIO_new_mem_buf(os->data, os->length);
+        } else if (os != NULL && os->length > 0) {
+            /*
+             * bio needs a copy of os->data instead of a pointer because
+             * the data will be used after os has been freed
+             */
+            bio = BIO_new(BIO_s_mem());
+            if (bio == NULL)
+                goto err;
+            if (BIO_write(bio, os->data, os->length) != os->length) {
+                BIO_free_all(bio);
+                goto err;
+            }
```

Comment:
```suggestion
            if (bio != NULL) {
                BIO_set_mem_eof_return(bio, 0);
                if (BIO_write(bio, os->data, os->length) != os->length) {
                    BIO_free_all(bio);
                    bio = NULL;
                }
            }
```
Memory BIOs default to signalling a retriable error at the end of the available data, rather than return EOF. I don't think that's what we want here. The caller should see a hard EOF after reading the entirety `os->data`.
See similar code in the `else` branch.

Reply:
Good point, it's been fixed.